pkg/controller/keyvault/spc/gateway.go

```diff
@@ -0,0 +1,204 @@
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/controllername"
+	"github.com/Azure/aks-app-routing-operator/pkg/controller/metrics"
+	"github.com/Azure/aks-app-routing-operator/pkg/util"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	gatewayv1 "sigs.k8s.io/gateway-api/apis/v1"
+	secv1 "sigs.k8s.io/secrets-store-csi-driver/apis/v1"
+)
+
+var gatewaySecretProviderControllerName = controllername.New("gateway", "keyvault", "secret", "provider")
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config, serviceAccountIndexName string) error {
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+	if conf.DisableKeyvault {
+		return nil
+	}
+
+	spcReconciler := &secretProviderClassReconciler[*gatewayv1.Gateway]{
+		name: gatewaySecretProviderControllerName,
+		toSpcOpts: func(ctx context.Context, cl client.Client, gw *gatewayv1.Gateway) iter.Seq2[spcOpts, error] {
+			return gatewayToSpcOpts(ctx, cl, conf, gw)
+		},
+
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	}
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}).
+			Owns(&secv1.SecretProviderClass{}).
+			Watches(&corev1.ServiceAccount{}, handler.EnqueueRequestsFromMapFunc(util.GenerateGatewayGetter(manager, serviceAccountIndexName))),
+		manager.GetLogger(),
+	).Complete(spcReconciler)
+}
+
+func gatewayToSpcOpts(ctx context.Context, cl client.Client, conf *config.Config, gw *gatewayv1.Gateway) iter.Seq2[spcOpts, error] {
+	return func(yield func(spcOpts, error) bool) {
+		if conf == nil {
+			yield(spcOpts{}, errors.New("config is nil"))
+			return
+		}
+
+		if gw == nil {
+			yield(spcOpts{}, errors.New("gateway is nil"))
+			return
+		}
+
+		if !IsManagedGateway(gw) {
+			// todo: test this and make sure it returns no values instead of hangs
+			return
+		}
+
+		for index, listener := range gw.Spec.Listeners {
+			name := GetGatewayListenerSpcName(gw.Name, string(listener.Name))
+			opts := spcOpts{
+				action:     actionReconcile,
+				name:       name,
+				namespace:  gw.Namespace,
+				tenantId:   conf.TenantID,
+				secretName: name,
+				cloud:      conf.Cloud,
+			}
+
+			if !ListenerIsKvEnabled(listener) {
+				opts.action = actionCleanup
+				if !yield(opts, nil) {
+					return
+				}
+				continue
+			}
+
+			clientId, err := clientIdFromListener(ctx, cl, gw.Namespace, listener)
+			if err != nil {
+				if !yield(opts, err) {
+					return
+				}
+				continue
+			}
+			opts.clientId = clientId
+
+			uri := string(listener.TLS.Options[certUriTLSOption])
+			certRef, err := parseKeyVaultCertURI(uri)
+			if err != nil {
+				if !yield(opts, fmt.Errorf("parsing KeyVault cert URI %s: %w", uri, err)) {
```

Comment:
do we not want to bubble this up as a user error? Won't fmt.Errorf overwrite that?

Reply:
good catch, fixed